### PR TITLE
ide: improve ReferenceCategoryType

### PR DESCRIPTION
**Pull request by matklad**
It is bitset semantically --- many categorical things can be true about a reference at the same time.

In parciular, a reference can be a "test" and a "write" at the same time.

### CODE REVIEW

File: crates/rust-analyzer/src/lsp/to_proto.rs [hunk 1]
```diff
@@ -92,12 +92,13 @@ pub(crate) fn structure_node_kind(kind: StructureNodeKind) -> lsp_types::SymbolK
 pub(crate) fn document_highlight_kind(
     category: ReferenceCategory,
 ) -> Option<lsp_types::DocumentHighlightKind> {
-    match category {
-        ReferenceCategory::Read => Some(lsp_types::DocumentHighlightKind::READ),
-        ReferenceCategory::Write => Some(lsp_types::DocumentHighlightKind::WRITE),
-        ReferenceCategory::Import => None,
-        ReferenceCategory::Test => None,
+    if category.contains(ReferenceCategory::READ) {
```
[the remark this answers is not in the document]
Comment (matklad) [reply crates/rust-analyzer/src/lsp/to_proto.rs]: Good call! 

>I'm not even sure how these work.

They just default to read, yeah, `S{f:0}` is a `Create`, which is another pre-existing todo!

[the remark this answers is not in the document]
Comment (lnicola) [reply crates/rust-analyzer/src/lsp/to_proto.rs]: Oh, of course that's what `retain_adt_literal_usages` doest :facepalm:.

**Review by lnicola — APPROVED** [review]

**Comment by matklad** [comment]
@lnicola did you mean to r+?

**Pull request merged**